File: part3/perf_course/src/repetition_tester.rs
pub mod repetition_tester {
    use crate::perf_metrics::{get_page_faults, high_resolution_info, high_resolution_time};
    use mach::mach_time::mach_timebase_info;
    use std::io::Write;
    use std::time::Duration;

    #[derive(PartialEq, Copy, Debug, Clone)]
    enum State {
        Uninitialized,
        Testing,
        Completed,
        Error,
    }

    #[derive(Debug, Copy, Clone)]
    enum RepetitionTesterMetrics {
        TestCount,
        Time,
        PageFaults,
        ByteCount,

        Count,
    }

    #[derive(Debug, Copy, Clone)]
    pub struct RepetitionTesterResults {
        pub totals: [u64; RepetitionTesterMetrics::Count as usize],
        pub min: [u64; RepetitionTesterMetrics::Count as usize],
        pub max: [u64; RepetitionTesterMetrics::Count as usize],
    }

    #[derive(Debug, Copy, Clone)]
    pub struct RepetitionTester {
        state: State,
        start_time: u64,
        time_to_wait: u64,
        cpu_timebase_info: mach_timebase_info,

        open_blocks_count: u32,
        close_blocks_count: u32,

        expected_bytes: u64,
        test_metrics: [u64; RepetitionTesterMetrics::Count as usize],

        results: RepetitionTesterResults,

        pid: i32,
    }

    impl RepetitionTester {
        pub fn new() -> Self {
            RepetitionTester {
                state: State::Uninitialized,
                start_time: 0,
                time_to_wait: 0,
                cpu_timebase_info: high_resolution_info(),
                open_blocks_count: 0,
                close_blocks_count: 0,
                expected_bytes: 0,
                pid: std::process::id() as i32,

                test_metrics: [0; RepetitionTesterMetrics::Count as usize],
                results: RepetitionTesterResults {
                    totals: [0; RepetitionTesterMetrics::Count as usize],
                    min: [0; RepetitionTesterMetrics::Count as usize],
                    max: [0; RepetitionTesterMetrics::Count as usize],
                },
            }
        }

        pub fn start_test_wave(&mut self, seconds_to_try: u64, expected_bytes: u64) {
            match self.state {
                State::Uninitialized => {
                    self.state = State::Testing;
                    self.start_time = high_resolution_time();
                    self.open_blocks_count = 0;
                    self.close_blocks_count = 0;
                    self.expected_bytes = expected_bytes;
                    self.results = RepetitionTesterResults {
                        totals: [0; RepetitionTesterMetrics::Count as usize],
                        min: [0; RepetitionTesterMetrics::Count as usize],
                        max: [0; RepetitionTesterMetrics::Count as usize],
                    };
                }
                State::Completed => {
                    self.state = State::Testing;
                }
                State::Testing => {
                    self.set_error("Test wave already in progress");
                }
                State::Error => {
                    self.set_error("Cannot start test wave from error state");
                }
            }

            self.start_time = high_resolution_time();
            self.time_to_wait = self.time_from_seconds(seconds_to_try);
        }

        pub fn set_error(&mut self, error: &str) {
            self.state = State::Error;
            eprintln!("[RepetitionTester] Error: {}", error);
        }

        pub fn begin_time(&mut self) {
            self.open_blocks_count += 1;
            self.test_metrics[RepetitionTesterMetrics::Time as usize] = high_resolution_time();
            self.test_metrics[RepetitionTesterMetrics::PageFaults as usize] =
                get_page_faults(self.pid) as u64;
        }

        pub fn end_time(&mut self) {
            self.close_blocks_count += 1;
            self.test_metrics[RepetitionTesterMetrics::Time as usize] =
                high_resolution_time() - self.test_metrics[RepetitionTesterMetrics::Time as usize];
            self.test_metrics[RepetitionTesterMetrics::PageFaults as usize] =
                get_page_faults(self.pid) as u64
                    - self.test_metrics[RepetitionTesterMetrics::PageFaults as usize];
        }

        pub fn count_bytes(&mut self, bytes: u64) {
            self.test_metrics[RepetitionTesterMetrics::ByteCount as usize] += bytes;
        }

        pub fn is_testing(&mut self) -> bool {
            if self.state == State::Testing {
                let current_time = high_resolution_time();
                if self.open_blocks_count > 0 {
                    if self.open_blocks_count != self.close_blocks_count {
                        self.set_error("Open and close blocks do not match");
                    }

                    if self.test_metrics[RepetitionTesterMetrics::ByteCount as usize]
                        != self.expected_bytes
                    {
                        self.set_error("Byte count does not match expected bytes");
                    }

                    if self.state == State::Testing {
                        self.test_metrics[RepetitionTesterMetrics::TestCount as usize] = 1;

                        for i in 0..RepetitionTesterMetrics::Count as usize {
                            self.results.totals[i] += self.test_metrics[i];
                        }

                        if self.results.max[RepetitionTesterMetrics::Time as usize]
                            < self.test_metrics[RepetitionTesterMetrics::Time as usize]
                        {
                            self.results.max = self.test_metrics;
                        }

                        if self.results.min[RepetitionTesterMetrics::Time as usize] == 0
                            || self.results.min[RepetitionTesterMetrics::Time as usize]
                                > self.test_metrics[RepetitionTesterMetrics::Time as usize]
                        {
                            self.results.min = self.test_metrics;

                            self.start_time = current_time;
                            self.print_time("Min", self.results.min, true);
                        }

                        self.open_blocks_count = 0;
                        self.close_blocks_count = 0;
                        self.test_metrics = [0; RepetitionTesterMetrics::Count as usize];
                    }

                    if current_time - self.start_time > self.time_to_wait {
                        self.state = State::Completed;
                        self.print_results();
                    }
                }
            }

            self.state == State::Testing
        }

        // helper functions
        fn time_from_seconds(&mut self, seconds: u64) -> u64 {
            seconds * 1_000_000_00 * self.cpu_timebase_info.denom as u64
                / self.cpu_timebase_info.denom as u64
        }

        fn time_as_seconds(&mut self, time: u64) -> Duration {
            Duration::from_nanos(
                time * self.cpu_timebase_info.numer as u64 / self.cpu_timebase_info.denom as u64,
            )
        }

        // fn print_time(&mut self, label: &str, time: u64, bytes: u64, carrier_return: bool) {
        fn print_time(
            &mut self,
            label: &str,
            value: [u64; RepetitionTesterMetrics::Count as usize],
            carrier_return: bool,
        ) {
            let test_count = value[RepetitionTesterMetrics::TestCount as usize];
            let mut local_value = [0; RepetitionTesterMetrics::Count as usize];
            for i in 0..RepetitionTesterMetrics::Count as usize {
                local_value[i] = value[i] / test_count as u64;
            }

            let time = local_value[RepetitionTesterMetrics::Time as usize];
            let time_in_seconds = self.time_as_seconds(time);
            // println!("{}: {} ({}s)", label, time, time_in_seconds.as_secs_f64());

            print!("{}: {} ({}s)", label, time, time_in_seconds.as_secs_f64());

            let bytes = local_value[RepetitionTesterMetrics::ByteCount as usize];
            if bytes > 0 {
                let gb_processed = bytes as f64 / (1024.0 * 1024.0 * 1024.0);
                let bandwidth = gb_processed / time_in_seconds.as_secs_f64();
                print!(" ({:.10} GB/s)", bandwidth);
            }

            let page_faults = local_value[RepetitionTesterMetrics::PageFaults as usize];
            if page_faults > 0 {
                print!(
                    " (PF: {:.4}, {:.4}k/fault)",
                    page_faults,
                    bytes as f64 / (page_faults as f64 * 1024.0)
                );
            }

            if carrier_return {
                print!("                   \r");
                std::io::stdout().flush().unwrap();
            } else {
                println!();
            }
        }

        fn print_results(&mut self) {
            self.print_time("Min", self.results.min, false);
            self.print_time("Max", self.results.max, false);
            self.print_time("Avg.", self.results.totals, false);
        }
    }
}

File: part5/src/repetition-tester.zig
const std = @import("std");
const debug = std.debug;
const perf = @import("perf-metrics.zig");

const State = enum {
    unitialized,
    testing,
    completed,
    err,
};

pub const Metrics = enum {
    testCount,
    time,
    pageFaults,
    byteCount,
    gbPerSec,
    kbPerPgFault,

    count,
};

pub const ResultValue = struct {
    values: [@intFromEnum(Metrics.count)]u64,
    perCount: [@intFromEnum(Metrics.count)]f64,
};

pub const Results = struct {
    totals: ResultValue,
    min: ResultValue,
    max: ResultValue,
};

fn timeAsSeconds(time: f64) f64 {
    return time / @as(f64, @floatFromInt(std.time.ns_per_s));
}

fn timeFromSecs(seconds: u64) u64 {
    return seconds * std.time.ns_per_s;
}

fn computeDerivedValues(value: *ResultValue) void {
    const testCount: f64 = @floatFromInt(value.values[@intFromEnum(Metrics.testCount)]);

    for (0..@intFromEnum(Metrics.count)) |i| {
        value.perCount[i] = @as(f64, @floatFromInt(value.values[i])) / testCount;
    }

    const time = timeAsSeconds(value.perCount[@intFromEnum(Metrics.time)]);
    const bytes = value.perCount[@intFromEnum(Metrics.byteCount)];

    if (bytes > 0) {
        const GB: f64 = 1024.0 * 1024.0 * 1024.0;
        const gbProcessed = bytes / (GB * time);
        value.perCount[@intFromEnum(Metrics.gbPerSec)] = gbProcessed;
    }

    const pageFaults = value.perCount[@intFromEnum(Metrics.pageFaults)];
    if (pageFaults > 0) {
        value.perCount[@intFromEnum(Metrics.kbPerPgFault)] = bytes / pageFaults * 1024;
    }
}

const MAX_LABEL_SIZE = 64;

pub fn ReptitionTestSeries(comptime maxRows: usize, comptime columnCount: usize) type {
    const maxResultsSize = maxRows * columnCount;
    return struct {
        maxRowCount: usize,
        columnCount: usize,
        rowIndex: usize,
        columnIndex: usize,
        currentRun: usize,
        rowLabels: [maxRows][MAX_LABEL_SIZE]u8, // Changed to array
        columnsLabel: [columnCount][MAX_LABEL_SIZE]u8, // Changed to array

        results: [maxResultsSize]Results,

        const Self = @This();

        pub fn new() Self {
            const size = @intFromEnum(Metrics.count);
            return Self{
                .maxRowCount = maxRows,
                .columnCount = columnCount,
                .rowIndex = 0,
                .columnIndex = 0,
                .currentRun = 0,
                .results = [_]Results{Results{
                    .totals = ResultValue{
                        .values = [_]u64{0} ** size,
                        .perCount = [_]f64{0} ** size,
                    },
                    .min = ResultValue{
                        .values = [_]u64{0} ** size,
                        .perCount = [_]f64{0} ** size,
                    },
                    .max = ResultValue{
                        .values = [_]u64{0} ** size,
                        .perCount = [_]f64{0} ** size,
                    },
                }} ** maxResultsSize,
                .rowLabels = [_][MAX_LABEL_SIZE]u8{[_]u8{0} ** MAX_LABEL_SIZE} ** maxRows,
                .columnsLabel = [_][MAX_LABEL_SIZE]u8{[_]u8{0} ** MAX_LABEL_SIZE} ** columnCount,
            };
        }

        pub fn isInbound(self: *Self) bool {
            return (self.rowIndex < self.maxRowCount) and (self.columnIndex < self.columnCount);
        }

        pub fn setRowLabel(self: *Self, label: []const u8) void {
            if (self.isInbound()) {
                self.rowLabels[self.rowIndex] = [_]u8{0} ** MAX_LABEL_SIZE;
                const copyLen = @min(label.len, MAX_LABEL_SIZE);
                @memcpy(self.rowLabels[self.rowIndex][0..copyLen], label[0..copyLen]);
            }
        }

        pub fn setColumnLabel(self: *Self, label: []const u8) void {
            if (self.isInbound()) {
                self.columnsLabel[self.columnIndex] = [_]u8{0} ** MAX_LABEL_SIZE;
                const copyLen = @min(label.len, MAX_LABEL_SIZE);
                @memcpy(self.columnsLabel[self.columnIndex][0..copyLen], label[0..copyLen]);
            }
        }

        pub fn newTestWave(self: *Self, t: *Tester, secondsToTry: u64, expectedBytes: u64) void {
            if (self.isInbound()) {
                const c = self.columnsLabel[self.columnIndex];
                const cl = std.mem.indexOfScalar(u8, &c, 0).?;
                const r = self.rowLabels[self.rowIndex];
                const rl = std.mem.indexOfScalar(u8, &r, 0).?;
                std.debug.print("\n--- {s} {s} ---\n", .{ c[0..cl], r[0..rl] });

                self.currentRun += 1;
            }
            t.startNewWave(secondsToTry, expectedBytes);
        }
        pub fn isTesting(self: *Self, t: *Tester) bool {
            const res = t.isTesting();
            if (!res) {
                if (self.isInbound()) {
                    // todo get results...
                    self.results[self.rowIndex * self.columnCount + self.columnIndex] = t.results;
                    // debug.print("=>=>=>=> rowIndex: {d}, columnCount: {d}, columnIndex: {d}, LOC: {d}\n", .{ self.rowIndex, self.columnCount, self.columnIndex, self.rowIndex * self.columnCount + self.columnIndex });

                    self.columnIndex += 1;
                    if (self.columnIndex >= self.columnCount) {
                        self.columnIndex = 0;
                        self.rowIndex += 1;
                    }
                }
            }

            return res;
        }
        pub fn dumpCSV(self: *Self, metric: Metrics) void {
            for (0..self.columnCount) |c| {
                const r = self.columnsLabel[c];
                const rl = std.mem.indexOfScalar(u8, &r, 0).?;
                debug.print("{s},", .{r[0..rl]});
            }
            debug.print("\n", .{});

            for (0..self.rowIndex + 1) |i| {
                for (0..self.columnCount) |c| {
                    // debug.print("->->->-> rowIndex: {d}, columnCount: {d}, columnIndex: {d}, LOC: {d}\n", .{ i, self.columnCount, c, i * self.columnCount + c });
                    const r = self.results[i * self.columnCount + c];
                    debug.print("{d},", .{r.min.perCount[@intFromEnum(metric)]});
                }
                debug.print("\n", .{});
            }
        }
    };
}

pub const Tester = struct {
    state: State,
    startTime: u64,
    timeToWait: u64,
    timeBaseInfo: perf.TBInfo,

    openBlocksCount: u32,
    closeBlockCount: u32,

    expectedBytes: u64,
    testMetrics: [@intFromEnum(Metrics.count)]u64,

    labelTimeBuffer: [64]u8,
    results: Results,

    pub fn new() Tester {
        const size = @intFromEnum(Metrics.count);
        return Tester{
            .state = State.unitialized,
            .startTime = 0,
            .timeToWait = 0,
            .timeBaseInfo = perf.timeBaseInfo(),
            .openBlocksCount = 0,
            .closeBlockCount = 0,
            .expectedBytes = 0,
            .testMetrics = [_]u64{0} ** size,
            .labelTimeBuffer = [_]u8{0} ** 64,
            .results = Results{
                .totals = ResultValue{
                    .values = [_]u64{0} ** size,
                    .perCount = [_]f64{0} ** size,
                },
                .min = ResultValue{
                    .values = [_]u64{0} ** size,
                    .perCount = [_]f64{0} ** size,
                },
                .max = ResultValue{
                    .values = [_]u64{0} ** size,
                    .perCount = [_]f64{0} ** size,
                },
            },
        };
    }

    pub fn startNewWave(self: *Tester, secondsToTry: u64, expectedBytes: u64) void {
        // debug.print(">>>> secondsToTry: {d}\n", .{secondsToTry});
        switch (self.state) {
            State.unitialized => {
                self.state = State.testing;
                self.startTime = perf.highResolutionClock();
                self.openBlocksCount = 0;
                self.closeBlockCount = 0;
                self.expectedBytes = expectedBytes;
                self.results = Results{
                    .totals = ResultValue{
                        .values = [_]u64{0} ** @intFromEnum(Metrics.count),
                        .perCount = [_]f64{0} ** @intFromEnum(Metrics.count),
                    },
                    .min = ResultValue{
                        .values = [_]u64{0} ** @intFromEnum(Metrics.count),
                        .perCount = [_]f64{0} ** @intFromEnum(Metrics.count),
                    },
                    .max = ResultValue{
                        .values = [_]u64{0} ** @intFromEnum(Metrics.count),
                        .perCount = [_]f64{0} ** @intFromEnum(Metrics.count),
                    },
                };

                const faults: u64 = @intCast(perf.getPageFaults());
                self.testMetrics[@intFromEnum(Metrics.pageFaults)] = faults - self.testMetrics[@intFromEnum(Metrics.pageFaults)];
            },
            State.completed => {
                self.state = State.testing;
            },
            State.testing => unreachable,
            State.err => unreachable,
        }

        self.startTime = perf.highResolutionClock();
        self.timeToWait = timeFromSecs(secondsToTry);
    }

    pub fn setError(self: *Tester, msg: []const u8) void {
        self.state = State.err;
        debug.print("{s}", .{msg});
    }

    pub fn beginTime(self: *Tester) void {
        self.openBlocksCount += 1;
        self.testMetrics[@intFromEnum(Metrics.time)] = perf.highResolutionClock();
        self.testMetrics[@intFromEnum(Metrics.pageFaults)] = @intCast(perf.getPageFaults());
        self.testMetrics[@intFromEnum(Metrics.byteCount)] = 0;
    }

    pub fn endTime(self: *Tester) void {
        self.closeBlockCount += 1;
        self.testMetrics[@intFromEnum(Metrics.time)] = perf.highResolutionClock() - self.testMetrics[@intFromEnum(Metrics.time)];
        const faults: u64 = @intCast(perf.getPageFaults());
        self.testMetrics[@intFromEnum(Metrics.pageFaults)] = faults - self.testMetrics[@intFromEnum(Metrics.pageFaults)];
    }

    pub fn countBytes(self: *Tester, bytes: u64) void {
        self.testMetrics[@intFromEnum(Metrics.byteCount)] += bytes;
    }

    pub fn isTesting(self: *Tester) bool {
        if (self.state == State.testing) {
            const currentTime = perf.highResolutionClock();
            if (self.openBlocksCount > 0) {
                if (self.openBlocksCount != self.closeBlockCount) {
                    self.setError("Open and close blocks count doesn't match\n");
                }

                if (self.testMetrics[@intFromEnum(Metrics.byteCount)] != self.expectedBytes) {
                    std.debug.print("expected: {d}, found: {d}\n", .{ self.expectedBytes, self.testMetrics[@intFromEnum(Metrics.byteCount)] });
                    self.setError("byteCount doesn't match\n");
                }

                if (self.state == State.testing) {
                    self.testMetrics[@intFromEnum(Metrics.testCount)] = 1;

                    for (0..@intFromEnum(Metrics.count)) |i| {
                        self.results.totals.values[i] += self.testMetrics[i];
                    }

                    if (self.results.max.values[@intFromEnum(Metrics.time)] < self.testMetrics[@intFromEnum(Metrics.time)]) {
                        self.results.max.values = self.testMetrics;
                    }

                    const rmin = self.results.min.values[@intFromEnum(Metrics.time)];
                    if (rmin == 0 or rmin > self.testMetrics[@intFromEnum(Metrics.time)]) {
                        self.results.min.values = self.testMetrics;
                        self.startTime = currentTime;
                        self.printTime("Min", self.results.min, true);
                    }
                }

                if (currentTime - self.startTime > self.timeToWait) {
                    self.state = State.completed;
                    computeDerivedValues(&self.results.totals);
                    computeDerivedValues(&self.results.min);
                    computeDerivedValues(&self.results.max);
                    self.printAllResults();
                }
            }

            return true;
        }

        return false;
    }

    fn printTime(self: *Tester, label: []const u8, value: ResultValue, carridgeReturn: bool) void {
        const time = value.values[@intFromEnum(Metrics.time)];
        var w: std.io.Writer = .fixed(&self.labelTimeBuffer);
        w.printDuration(time, .{}) catch unreachable;
        debug.print("{s}: {} ({s})", .{ label, time, w.buffered() });

        const bytes = value.perCount[@intFromEnum(Metrics.byteCount)];
        if (bytes > 0) {
            debug.print(" ({d:.10} GB/s)", .{value.perCount[@intFromEnum(Metrics.gbPerSec)]});
        }

        const pageFaults = value.perCount[@intFromEnum(Metrics.pageFaults)];
        if (pageFaults > 0) {
            debug.print(" (PF: {d:.4}, {d:.4}k/faults)", .{ pageFaults, value.perCount[@intFromEnum(Metrics.kbPerPgFault)] });
        }

        if (carridgeReturn) {
            debug.print("               \r", .{});
        } else {
            debug.print("\n", .{});
        }
    }

    pub fn printAllResults(self: *Tester) void {
        self.printTime("Min", self.results.min, false);
        self.printTime("Max", self.results.max, false);
        self.printTime("Avg.", self.results.totals, false);
    }
};

fn run(pageSize: usize, pageCount: usize) !void {
    const totalSize = pageSize * pageCount;

    const prot = std.posix.PROT.READ | std.posix.PROT.WRITE;

    for (0..pageCount) |touchCount| {
        const touchSize = pageSize * touchCount;
        const addr = try std.posix.mmap(null, totalSize, prot, .{ .TYPE = .PRIVATE, .ANONYMOUS = true }, -1, 0);
        defer std.posix.munmap(addr);

        var ptr: [*]u8 = @ptrCast(addr);
        // const startFaults = perf.getPageFaults();
        for (0..touchSize) |i| {
            ptr[i] = @truncate(i);
        }
        // const endFaults = perf.getPageFaults();
        // _ = endFaults - startFaults;
        // const faultCount = endFaults - startFaults;
        // debug.print("{d}, {d}, {d}\n", .{ pageCount, touchCount, faultCount });
    }
}

pub fn main() !void {
    // var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    // _ = gpa.allocator();

    var ts = ReptitionTestSeries(5, 2).new();
    ts.setRowLabel("row"[0..]);
    // std.debug.print(">>> {any} \n", .{ts});

    for (0..2) |i| {
        var buf: [64]u8 = [_]u8{0} ** 64;
        _ = try std.fmt.bufPrint(&buf, "column {d}", .{i});
        ts.setColumnLabel(buf[0..]);

        var t = Tester.new();
        // while (true) {
        // t.startNewWave(5, 10);
        ts.newTestWave(&t, 5, 10 * 1024 * 1024);

        while (ts.isTesting(&t)) {
            t.beginTime();
            try run(4096, 128);
            t.endTime();
            t.countBytes(10 * 1024 * 1024);
        }
    }
    ts.dumpCSV(Metrics.gbPerSec);
    // }
    // t.startNewWave(2, 10000);
    // t.beginTime();
    // debug.print("hello world >>> {any}\n", .{t});
    // t.endTime();
    // t.countBytes(10000);
    // debug.print("hello world >>> {any}\n", .{t});

}
